pkg/database/resync.go

```diff
@@ -3,60 +3,162 @@
 package database
 
 import (
+	"bytes"
 	"context"
 	"encoding/json"
 	"fmt"
+	"io"
 	"reflect"
+	"runtime"
+	"strings"
 	"time"
 
 	"github.com/stolostron/search-indexer/pkg/metrics"
 	"github.com/stolostron/search-indexer/pkg/model"
 	"k8s.io/klog/v2"
 )
 
+func PrintMem(msg string) {
+	fmt.Printf(msg + strings.Repeat("\t", 6-len(msg)/8))
+	bToMb := func(b uint64) uint64 {
+		return b / 1024 / 1024
+	}
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+	// For info on each, see: https://golang.org/pkg/runtime/#MemStats
+	fmt.Printf("Alloc = %v MiB", bToMb(m.Alloc))
+	// fmt.Printf("\tTotalAlloc = %v MiB", bToMb(m.TotalAlloc))
+	// fmt.Printf("\tSys = %v MiB", bToMb(m.Sys))
+	fmt.Printf("\tobjects: %v", m.HeapObjects)
+	fmt.Printf("\tNumGC = %v\n", m.NumGC)
+
+	runtime.GC()
+	runtime.ReadMemStats(&m)
+	fmt.Printf(msg + " (GC)" + strings.Repeat("\t", 6-len(msg+" (GC)")/8))
+	fmt.Printf("Alloc = %v MiB", bToMb(m.Alloc))
+	fmt.Printf("\tobjects: %v", m.HeapObjects)
+	fmt.Printf("\tNumGC = %v\n", m.NumGC)
+}
+
 // Reset data for the cluster to the incoming state.
 func (dao *DAO) ResyncData(ctx context.Context, event model.SyncEvent,
-	clusterName string, syncResponse *model.SyncResponse) error {
+	clusterName string, syncResponse *model.SyncResponse, body []byte) error {
+	PrintMem("ResyncData start")
 
 	defer metrics.SlowLog(fmt.Sprintf("Slow resync from %12s. RequestId: %d", clusterName, event.RequestId), 0)()
 	klog.Infof(
 		"Starting resync from %12s. This is normal, but it could be a problem if it happens often.", clusterName)
 
 	// Reset resources
-	err := dao.resetResources(ctx, event.AddResources, clusterName, syncResponse)
+	PrintMem("resetResources start")
+	err := dao.resetResources(ctx, clusterName, syncResponse, body)
 	if err != nil {
 		klog.Warningf("Error resyncing resources for cluster %12s. Error: %+v", clusterName, err)
 		return err
 	}
+	PrintMem("resetResources end")
 
+	PrintMem("resetEdges start")
 	// Reset edges
-	err = dao.resetEdges(ctx, event.AddEdges, clusterName, syncResponse)
+	err = dao.resetEdges(ctx, clusterName, syncResponse, body)
 	if err != nil {
 		klog.Warningf("Error resyncing edges for cluster %12s. Error: %+v", clusterName, err)
 		return err
 	}
+	PrintMem("resetEdges end")
 
 	klog.V(1).Infof("Completed resync of cluster %12s.\t RequestId: %d", clusterName, event.RequestId)
+	PrintMem("ResyncData end")
 	return nil
 }
 
+func decodeAddResources(body []byte, key string) (map[string]*model.Resource, int, error) {
```

Comment:
This decodes all resources at once, so we are using memory until the resources are processed.

Reply:
We have 2 options.
1. We process each resource from here. That means finding a way to call batch.Queue() from here.
2. Or we can keep the []byte of each resource and decode it later.